File: asr-file-ws.py
#!/usr/bin/env python3
"""
Python WebSocket client for Voxist ASR streaming
"""

import asyncio
import contextlib
import json
import os
import sys
import time
import websockets
from pathlib import Path
from urllib.parse import quote

from wav import WavError, assert_streamable, read_wav_info

# Only 16 kHz is served by the streaming engines. The gateway does not resample,
# so this is not a knob - it is a property of the audio you must supply.
SAMPLE_RATE = 16000
BYTES_PER_SAMPLE = 2
CHUNK_DURATION_MS = 100

# If the gateway's upstream engine socket is not yet OPEN when `Done` arrives,
# the gateway drops the flush and never closes us, so waiting for the close
# frame unconditionally would hang forever.
DRAIN_TIMEOUT_S = 30

# 1005 is "no status received". Neither it nor 1000 indicates a problem. Other
# codes may be relayed verbatim from the upstream ASR engine, so they are only
# treated as failures when no final result arrived.
CLEAN_CLOSE_CODES = (1000, 1005)


class ASRWebSocketClient:
    def __init__(self, wav_file_path, api_key, lang="fr", is_staging=False, punctuation_mode=None):
        self.wav_file_path = Path(wav_file_path)
        self.api_key = api_key
        self.lang = lang
        self.sample_rate = SAMPLE_RATE
        self.is_staging = is_staging
        self.punctuation_mode = punctuation_mode
        self.wav_info = None

        # Select the appropriate domain based on staging flag.
        # VOXIST_ASR_URL overrides the base URL entirely (self-hosted gateway, local tests).
        self.domain = 'asr-staging-dev.voxist.com' if is_staging else 'api-asr.voxist.com'
        self.base_url = os.environ.get('VOXIST_ASR_URL', f"wss://{self.domain}")
        self.url = (
            f"{self.base_url}/ws"
            f"?api_key={quote(api_key)}&lang={quote(lang)}&sample_rate={self.sample_rate}"
        )
        if punctuation_mode:
            self.url += f"&punctuation_mode={quote(punctuation_mode)}"

        self.CHUNK_SIZE = int(self.sample_rate * BYTES_PER_SAMPLE * (CHUNK_DURATION_MS / 1000))

        # State tracking
        self.start_time = None
        self.first_word_received = False
        self.last_segment = None
        self.received_final = False

    def clear_line(self):
        """Clear current line and move cursor to beginning"""
        print('\r' + ' ' * 80 + '\r', end='', flush=True)

    def validate_audio(self):
        """Parse the container and reject audio the gateway would mis-transcribe.

        The gateway treats every binary frame as raw PCM: it does not resample,
        does not downmix, and does not inspect the WAV header. Sending 8 kHz or
        stereo audio therefore yields a wrong transcript rather than an error.
        """
        self.wav_info = read_wav_info(str(self.wav_file_path))
        assert_streamable(self.wav_info, self.sample_rate)

    async def send_audio_chunks(self, websocket):
        """Stream the `data` chunk in real-time-sized pieces.

        Reads by offset rather than through the stdlib `wave` module so the
        header never reaches the socket and a pipe-written file (declared data
        length 0) still streams the audio it actually contains.
        """
        self.start_time = time.time()

        try:
            with open(self.wav_file_path, 'rb') as f:
                f.seek(self.wav_info.data_offset)
                remaining = self.wav_info.data_size
                while remaining > 0:
                    chunk = f.read(min(self.CHUNK_SIZE, remaining))
                    if not chunk:
                        break
                    remaining -= len(chunk)

                    await websocket.send(chunk)
                    await asyncio.sleep(CHUNK_DURATION_MS / 1000)
        except OSError as e:
            # File-read failures only. A ConnectionClosed raised by send() must
            # propagate: it carries the close code the caller reports.
            print(f"Error reading audio file: {e}")
            return

        # End-of-stream signal. This MUST be the text frame `Done`: it is what
        # flushes the decoder's tail and promotes the last partial to a final.
        # The server closes the socket once the engine has drained.
        await websocket.send('Done')

    def handle_message(self, message_data):
        """Handle incoming WebSocket messages"""
        try:
            message = json.loads(message_data)

            # Not every server message is a transcript: `text` may be absent.
            text = message.get('text')
            if isinstance(text, str) and text != '':
                if not self.first_word_received:
                    self.first_word_received = True
                    elapsed = int((time.time() - self.start_time) * 1000)
                    print(f'First word: {elapsed} ms')

                if message.get('type') == 'partial':
                    # Clear line and show partial result (overwriting previous partial)
                    self.clear_line()
                    print(message['text'], end='', flush=True)

                elif message.get('type') == 'final':
                    self.received_final = True
                    # Clear the partial result and print final result on new line
                    self.clear_line()
                    current_segment = message.get('segment')
                    # Only print if it's a new segment
                    if current_segment != self.last_segment:
                        print(message['text'])
                        self.last_segment = current_segment

        except json.JSONDecodeError:
            print(f"Error: Could not parse message: {message_data}")
        except Exception as e:
            print(f"Error handling message: {e}")

    def report_close(self, exc):
        """Explain an abnormal close using the gateway's documented codes."""
        code = getattr(exc, 'code', None)
        reason = (getattr(exc, 'reason', '') or '').strip()

        detail = reason
        if reason.startswith('{'):
            try:
                payload = json.loads(reason)
                detail = payload.get('message', reason)
                if payload.get('retryAfterMs'):
                    detail += f" (retry after {payload['retryAfterMs']} ms)"
            except json.JSONDecodeError:
                pass

        explanations = {
            1008: ('Unsupported language code, a model this account is not entitled to, '
                   'or a per-tenant rate limit'),
            1011: 'ASR engine unavailable, timed out, or not configured for this language',
            1013: 'Server at capacity - back off and retry',
        }
        print(f"\nConnection closed by server: code {code}"
              f"{' - ' + detail if detail else ''}")
        if code in explanations:
            print(f"  {explanations[code]}")

    async def receive_messages(self, websocket):
        """Consume server messages until the connection closes."""
        async for message in websocket:
            self.handle_message(message)

    async def run(self):
        """Run the WebSocket client. Returns a process exit status."""
        print(f"Environment: {'Staging' if self.is_staging else 'Production'}")
        print(f"Connecting to: {self.base_url}/ws?api_key=***&lang={self.lang}&sample_rate={self.sample_rate}")
        print(f"Audio file: {self.wav_file_path}")
        print(f"Language: {self.lang}")
        print(f"Punctuation mode: {self.punctuation_mode or 'Generated (default)'}")
        print(f"Sample rate: {self.sample_rate} Hz")
        print(f"Chunk size: {self.CHUNK_SIZE} bytes")
        print('')

        exit_code = 0
        send_task = None
        recv_task = None

        try:
            async with websockets.connect(self.url) as websocket:
                print('Connected to WebSocket')

                # Receive concurrently with sending, so a close that arrives
                # mid-stream is observed immediately rather than after the last
                # chunk, and so the sender's own failures are not masked.
                recv_task = asyncio.create_task(self.receive_messages(websocket))
                send_task = asyncio.create_task(self.send_audio_chunks(websocket))

                await send_task

                # `Done` has been sent; the server closes once the engine has
                # drained. Bounded, because a dropped flush never closes.
                try:
                    await asyncio.wait_for(recv_task, timeout=DRAIN_TIMEOUT_S)
                except asyncio.TimeoutError:
                    print(f"\nNo response from server {DRAIN_TIMEOUT_S} s after Done - closing.")
                    print('The final segment may be missing.')
                    exit_code = 1

        except websockets.exceptions.ConnectionClosedOK:
            pass
        except websockets.exceptions.ConnectionClosed as e:
            # 1008 / 1011 / 1013 land here. Surfacing the code matters: without
            # it a capacity rejection is indistinguishable from a clean run.
            code = getattr(e, 'code', None)
            if code in CLEAN_CLOSE_CODES:
                pass
            elif code in (1008, 1011, 1013) or not self.received_final:
                self.report_close(e)
                exit_code = 1
            else:
                # Relayed engine code, but the transcript did arrive.
                self.report_close(e)
        except OSError as e:
            print(f"Connection error: {e}")
            exit_code = 1
        except Exception as e:
            print(f"WebSocket error: {e}")
            exit_code = 1
        finally:
            for task in (send_task, recv_task):
                if task is not None and not task.done():
                    task.cancel()
                    with contextlib.suppress(asyncio.CancelledError, Exception):
                        await task
            elapsed = int((time.time() - self.start_time) * 1000) if self.start_time else 0
            print(f'\nFinished: {elapsed} ms')

        return exit_code


def main():
    # Parse command line arguments
    args = sys.argv[1:]
    is_staging = '--staging' in args

    # Remove --staging from args if present
    if is_staging:
        args.remove('--staging')

    # Optional --punctuation-mode=Generated|Dictated
    punctuation_mode = None
    for i, arg in enumerate(list(args)):
        if arg.startswith('--punctuation-mode'):
            if '=' in arg:
                punctuation_mode = arg.split('=', 1)[1]
                args.pop(i)
            else:
                punctuation_mode = args[i + 1] if i + 1 < len(args) else None
                del args[i:i + 2]
            if punctuation_mode not in ('Generated', 'Dictated'):
                print(f"Error: --punctuation-mode must be Generated or Dictated (got '{punctuation_mode}')")
                sys.exit(1)
            break

    if len(args) < 2:
        print("Usage: python asr-file-ws.py <API_KEY> <WAV_FILE> [LANG] [--punctuation-mode=MODE] [--staging]")
        print("Example: python asr-file-ws.py your-api-key audio.wav fr")
        print("Example: python asr-file-ws.py your-staging-api-key audio.wav fr-medical --staging")
        print("")
        print("Parameters:")
        print("  API_KEY: Your Voxist API key")
        print("  WAV_FILE: Path to the WAV audio file (16 kHz mono 16-bit PCM)")
        print("  LANG: Language code (optional, default: 'fr')")
        print("  --punctuation-mode: Generated (default) or Dictated, for spoken punctuation")
        print("  --staging: Use staging environment (optional)")
        print("")
        print("Supported Languages:")
        print("  fr: French")
        print("  fr-medical: French Medical")
        print("  fr-medicalV2-16 / fr-medicalV2-32 / fr-medicalV2-64: French Medical latency tiers")
        print("  en: English")
        print("  de: German")
        print("  es: Spanish")
        print("  it: Italian")
        print("  nl: Dutch")
        print("  pt: Portuguese")
        print("")
        print("Environments:")
        print("  Production: api-asr.voxist.com (default)")
        print("  Staging: asr-staging-dev.voxist.com (with --staging flag)")
        print("")
        print("Note: Staging and production use different API keys")
        sys.exit(1)

    api_key = args[0]
    wav_file = args[1]
    lang = args[2] if len(args) > 2 else "fr"

    client = ASRWebSocketClient(wav_file, api_key, lang, is_staging, punctuation_mode)

    try:
        client.validate_audio()
    except FileNotFoundError:
        print(f"Error: Could not find audio file at {wav_file}")
        sys.exit(1)
    except (WavError, OSError) as e:
        print(f"Error: {e}")
        sys.exit(1)

    sys.exit(asyncio.run(client.run()))


if __name__ == "__main__":
    main()

File: asr-file-ws.js
import WebSocket from 'ws';
import fs from 'fs';
import { readWavInfo, assertStreamable } from './wav.js';
import { closeExitCode, describeClose, exitCleanly, parsePunctuationMode, shouldReportClose } from './cli.js';

// If the gateway's upstream engine socket is not yet OPEN when `Done` arrives,
// the gateway drops the flush and never closes us (gateway `Done` handler), so
// waiting for the close frame unconditionally would hang forever.
const DRAIN_TIMEOUT_MS = 30000;

// Only 16 kHz is served by the streaming engines. The gateway does not resample,
// so this is not a knob — it is a property of the audio you must supply.
const SAMPLE_RATE = 16000;
const BYTES_PER_SAMPLE = 2;
const CHUNK_DURATION_MS = 100;
const CHUNK_SIZE = Math.floor(SAMPLE_RATE * BYTES_PER_SAMPLE * (CHUNK_DURATION_MS / 1000));

// Parse command line arguments
const args = process.argv.slice(2);
const stagingIndex = args.indexOf('--staging');
const isStaging = stagingIndex !== -1;

// Remove --staging from args if present
if (isStaging) {
  args.splice(stagingIndex, 1);
}

// Optional --punctuation-mode=Generated|Dictated
const punctuationMode = parsePunctuationMode(args);

if (args.length < 2) {
  console.log('Usage: node asr-file-ws.js <API_KEY> <WAV_FILE> [LANG] [--punctuation-mode=MODE] [--staging]');
  console.log('Example: node asr-file-ws.js your-api-key audio.wav fr');
  console.log('Example: node asr-file-ws.js your-staging-api-key audio.wav fr-medical --staging');
  console.log('');
  console.log('Parameters:');
  console.log('  API_KEY: Your Voxist API key');
  console.log('  WAV_FILE: Path to the WAV audio file (16 kHz mono 16-bit PCM)');
  console.log('  LANG: Language code (optional, default: "fr")');
  console.log('  --punctuation-mode: Generated (default) or Dictated, for spoken punctuation');
  console.log('  --staging: Use staging environment (optional)');
  console.log('');
  console.log('Supported Languages:');
  console.log('  fr: French');
  console.log('  fr-medical: French Medical');
  console.log('  fr-medicalV2-16 / fr-medicalV2-32 / fr-medicalV2-64: French Medical latency tiers');
  console.log('  en: English');
  console.log('  de: German');
  console.log('  es: Spanish');
  console.log('  it: Italian');
  console.log('  nl: Dutch');
  console.log('  pt: Portuguese');
  console.log('');
  console.log('Environments:');
  console.log('  Production: api-asr.voxist.com (default)');
  console.log('  Staging: asr-staging-dev.voxist.com (with --staging flag)');
  console.log('');
  console.log('Note: Staging and production use different API keys');
  process.exit(1);
}

const apiKey = args[0];
const wavFilePath = args[1];
const lang = args[2] || 'fr';

// Validate file exists
if (!fs.existsSync(wavFilePath)) {
  console.error(`Error: Could not find audio file at ${wavFilePath}`);
  process.exit(1);
}

// Validate the audio up front. The gateway accepts any binary frame as PCM, so a
// stereo / 8 kHz / MP3-in-a-.wav file produces a plausible-looking but wrong
// transcript instead of an error.
let wavInfo;
try {
  wavInfo = readWavInfo(wavFilePath);
  assertStreamable(wavInfo, SAMPLE_RATE);
} catch (error) {
  console.error(`Error: ${error.message}`);
  process.exit(1);
}

// Select the appropriate domain based on staging flag.
// VOXIST_ASR_URL overrides the base URL entirely (self-hosted gateway, local tests).
const domain = isStaging ? 'asr-staging-dev.voxist.com' : 'api-asr.voxist.com';
const baseUrl = process.env.VOXIST_ASR_URL || `wss://${domain}`;
const url = `${baseUrl}/ws?api_key=${encodeURIComponent(apiKey)}&lang=${encodeURIComponent(
  lang,
)}&sample_rate=${SAMPLE_RATE}${punctuationMode ? `&punctuation_mode=${punctuationMode}` : ''}`;

console.log(`Environment: ${isStaging ? 'Staging' : 'Production'}`);
console.log(`Connecting to: ${baseUrl}/ws?api_key=***&lang=${lang}&sample_rate=${SAMPLE_RATE}`);
console.log(`Audio file: ${wavFilePath}`);
console.log(`Language: ${lang}`);
console.log(`Punctuation mode: ${punctuationMode || 'Generated (default)'}`);
console.log(`Sample rate: ${SAMPLE_RATE} Hz`);
console.log(`Chunk size: ${CHUNK_SIZE} bytes`);
console.log('');


const ws = new WebSocket(url);
let start = Date.now();
let first = true;
let lastSegment = null;
let receivedFinal = false;
let doneSent = false;
let drainTimedOut = false;
let readFailed = false;
let bytesSent = 0;
let drainTimer = null;

// Clear current line and move cursor to beginning.
// Guarded: these are TTY-only APIs and throw when stdout is a pipe or a file.
const clearLine = () => {
  if (process.stdout.isTTY) {
    process.stdout.clearLine(0);
    process.stdout.cursorTo(0);
  } else {
    process.stdout.write('\n');
  }
};

ws.on('open', async () => {
  console.log('Connected to WebSocket');

  // Skip the WAV header — stream the `data` chunk only. readWavInfo guarantees
  // dataSize > 0, so the range below is always valid; the guard keeps any
  // surprise from becoming an unhandled rejection inside this async handler.
  let readStream;
  try {
    readStream = fs.createReadStream(wavFilePath, {
      start: wavInfo.dataOffset,
      end: wavInfo.dataOffset + wavInfo.dataSize - 1,
      highWaterMark: CHUNK_SIZE,
    });
  } catch (error) {
    console.error(`Error opening audio data: ${error.message}`);
    ws.close(1000, 'Client aborted');
    process.exit(1);
  }
  start = Date.now();

  readStream.on('data', async (chunk) => {
    if (ws.readyState === WebSocket.OPEN) {
      readStream.pause();
      bytesSent += chunk.length;
      ws.send(chunk);
      await new Promise((resolve) => setTimeout(resolve, CHUNK_DURATION_MS));
      readStream.resume();
    }
  });

  readStream.on('end', () => {
    // A short read ends the stream normally, so without this check a file
    // truncated underneath us would flush, transcribe a fragment, and exit 0.
    if (bytesSent < wavInfo.dataSize) {
      console.error(
        `\nWarning: sent ${bytesSent} of ${wavInfo.dataSize} declared audio bytes ` +
          '- the file was shorter than its header claims.',
      );
      readFailed = true;
    }
    if (ws.readyState === WebSocket.OPEN) {
      // End-of-stream signal. This MUST be the text frame `Done`: it is what
      // flushes the decoder's tail and promotes the last partial to a final.
      // The server then closes the socket once the engine has drained, so we
      // wait for `close` rather than hanging up ourselves.
      ws.send('Done');
      doneSent = true;

      // Backstop: if the gateway dropped the flush (upstream engine still
      // connecting) it will never close us, and this would otherwise hang.
      drainTimer = setTimeout(() => {
        console.error(`\nNo response from server ${DRAIN_TIMEOUT_MS} ms after Done - closing.`);
        console.error('The final segment may be missing.');
        drainTimedOut = true;
        ws.close(1000, 'Client drain timeout');
      }, DRAIN_TIMEOUT_MS);
    }
  });

  readStream.on('error', (error) => {
    // A partial upload is a failed run. Close with an explicit code: a bare
    // ws.close() surfaces locally as 1005, which reads as a clean shutdown.
    console.error(`Error reading file: ${error.message}`);
    readFailed = true;
    ws.close(1000, 'Client read error');
  });
});

ws.on('message', (data) => {
  try {
    let message = JSON.parse(data);
    // Not every server message is a transcript: `text` may be absent entirely.
    if (typeof message.text === 'string' && message.text !== '') {
      if (first) {
        first = false;
        console.log('First word: ' + (Date.now() - start) + ' ms');
      }

      if (message.type === 'partial') {
        // Clear line and show partial result (overwriting previous partial)
        clearLine();
        process.stdout.write(message.text);
      } else if (message.type === 'final') {
        receivedFinal = true;
        // Clear the partial result and print final result on new line
        clearLine();
        // Only print if it's a new segment
        const currentSegment = message.segment;
        if (currentSegment !== lastSegment) {
          console.log(message.text);
          lastSegment = currentSegment;
        }
      }
    }
  } catch (error) {
    console.error(`Error parsing message: ${error.message}`);
  }
});

ws.on('error', (error) => {
  console.error(`WebSocket error: ${error.message}`);
  console.error(`Failed to connect to: ${baseUrl}/ws`);
  exitCleanly(1);
});

ws.on('close', (code, reason) => {
  if (drainTimer) clearTimeout(drainTimer);
  console.log('\nFinished: ' + (Date.now() - start) + ' ms');

  // A drain timeout or a read error closes with 1000, but the run did not
  // succeed, so neither can be inferred from the close code alone.
  const status =
    drainTimedOut || readFailed ? 1 : closeExitCode(code, { doneSent, receivedFinal });

  // Reporting the code matters: without it a capacity rejection or an
  // unsupported-language close is indistinguishable from a clean run.
  if (shouldReportClose(code, status)) describeClose(code, reason);
  if (status !== 0 && !doneSent && !readFailed) {
    console.log('  The audio was not fully sent - the transcript is incomplete.');
  }
  if (status !== 0 && !receivedFinal) console.log('  No final transcription was received.');
  exitCleanly(status);
});
